src/steps/generic.rs

```diff
@@ -712,13 +712,94 @@ pub fn run_dotnet_upgrade(ctx: &ExecutionContext) -> Result<()> {
 
     print_separator(".NET");
 
-    for package in packages {
-        let package_name = package.split_whitespace().next().unwrap();
-        ctx.run_type()
+    let installed_packages = packages.map(|l| {
+        let mut iter = l.split_whitespace();
+        let name = iter.next();
+        let version = iter.next();
+        (name.unwrap_or(""), version.unwrap_or(""))
+    }).filter(|(name, _)| !name.is_empty());
+
+    let mut packages: Vec<(&str, &str, &str, bool)> = vec![];
+
+    for (name, version) in installed_packages {
+        let s_output = match ctx.run_type()
             .execute(&dotnet)
-            .args(["tool", "update", package_name, "--global"])
-            .status_checked()
-            .with_context(|| format!("Failed to update .NET package {package_name}"))?;
+            .args(["tool", "search", name])
+            // dotnet will print a greeting message on its first run, from this question:
+            // https://stackoverflow.com/q/70493706/14092446
+            // Setting `DOTNET_NOLOGO` to `true` should disable it
+            .env("DOTNET_NOLOGO", "true")
+            .output_checked_utf8() {
+                Ok(s_output) => Some(s_output),
+                Err(_) => None,
+            };
+        
+        if let Some(s_output) = s_output {
+            let mut search_packages = s_output
+            .stdout
+            .lines()
+            // Skip the header:
+            //
+            // Package ID            Latest Version      Authors               Downloads      Verified
+            // ---------------------------------------------------------------------------------------
+            //
+            // One thing to note is that .NET SDK respect locale, which means this
+            // header can be printed in languages other than English, do NOT use it
+            // to do any check.
+            .skip(2)
```

Comment:
It seems that `dotnet search <PKG>` won't output this header if `<PKG>` is not found:

```sh
$ dotnet tool search rust
Could not find any results.
```

Will this happen if the package has already been installed?

Reply:
If an installed package listed by `dotnet tool list --global` is guaranteed to be found with `dotnet tool search <PKG>`, then we can use the `--take 1` option to find the exact match:

```sh
$ dotnet tool search dotnet-dump
Package ID                     Latest Version      Authors               Downloads      Verified
------------------------------------------------------------------------------------------------
dotnet-dump                    7.0.447801          Microsoft             21089065
dotnet-gcdump                  7.0.447801          Microsoft             9036932
dumpdiag                       0.10.0              DumpDiag              2427
netspark.tools.dumpconfig      1.0.0               Gennadii Kostrov      286

$ dotnet tool search dotnet-dump --take 1
Package ID       Latest Version      Authors        Downloads      Verified
---------------------------------------------------------------------------
dotnet-dump      7.0.447801          Microsoft      21089065
```

Reply:
It seems like dotnet tool doesn't take into account whether the tool version supports the installed toolchain or not.
For example for package `dotnet-ef`, a few days ago version `8.0.0` released, dropping support for pre-dotnet-8.0, but the cli still tried installing the update, and it's still showing as latest available using `dotnet tool search`. Even tho I only had 6.0 and 7.0 installed.